muxers/mplex/src/io.rs

```diff
@@ -346,37 +352,96 @@ where
         self.guard_open()?;
 
         // Try to read from the buffer first.
-        if let Some((pos, data)) = self.buffer.iter()
-            .enumerate()
-            .find_map(|(pos, frame)| match frame {
-                Frame::Data { stream_id, data }
-                    if stream_id.into_local() == id => Some((pos, data.clone())),
-                _ => None
-            })
-        {
-            if self.buffer.len() == self.config.max_buffer_len {
-                // The buffer is full and no longer will be, so notify all pending readers.
-                ArcWake::wake_by_ref(&self.notifier_read);
+        if let Some(state) = self.substreams.get_mut(&id) {
+            let buf = state.recv_buf();
+            if !buf.is_empty() {
+                if self.blocked_stream == Some(id) {
+                    // Unblock reading new frames.
+                    self.blocked_stream = None;
+                    ArcWake::wake_by_ref(&self.notifier_read);
+                }
+                let data = buf.remove(0);
+                return Poll::Ready(Ok(Some(data)))
             }
-            self.buffer.remove(pos);
-            return Poll::Ready(Ok(Some(data)));
+            // If the stream buffer "spilled" onto the heap, free that memory.
+            buf.shrink_to_fit();
         }
 
         loop {
             // Check if the targeted substream (if any) reached EOF.
             if !self.can_read(&id) {
+                // Note: Contrary to what is recommended by the spec, we must
+                // return "EOF" also when the stream has been reset by the
+                // remote, as the `StreamMuxer::read_substream` contract only
+                // permits errors on "terminal" conditions, e.g. if the connection
+                // has been closed or on protocol misbehaviour.
                 return Poll::Ready(Ok(None))
             }
 
+            // Check if there is a blocked stream.
+            if let Some(blocked_id) = &self.blocked_stream {
+                // We have a blocked stream and cannot continue reading
+                // new frames for any stream, until frames are taken from
+                // the blocked stream's buffer. Try to wake pending readers
```

Comment:
```suggestion
                // the blocked stream's buffer. Try to wake a pending reader
```

As far as I can tell we can only wake the most recent reader, right?

Reply:
Yes, that comment is a left-over from the previous implementation. Thanks!